### PR TITLE
Get head v1 request

### PR DESCRIPTION
**Pull request by HelclM**
Implementing GET and HEAD on `/api/v1/files` url according to  https://dev.prusa3d.com/browse/BFW-2787

### CODE REVIEW

File: lib/WUI/link_content/prusa_link_api.cpp [hunk 1]
```diff
@@ -33,6 +33,64 @@ namespace {
...
+             * or such URIs, so we don't remap anything else but the "root".
+             */
+            static const char *const roots[] = {
+                "",
+                "/",
+                "/local",
+                "/local/",
+                "/sdcard",
+                "/sdcard/",
+            };
```
Comment (vorner) [inline lib/WUI/link_content/prusa_link_api.cpp:71]: Oops. I know it was my suggestion to re-use the code in the `/api/files`. But looking at it here, we probably don't want this particular re-mapping hack in the `/api/v1/files` ‒ one reason for creating the new API was to get rid of hardcoded `sdcard` and `local` and all the hacks related to that.

So I'd say we either want to somehow pass the list of mappings in as a parameter, or disable this part by some parameter.

File: lib/WUI/link_content/prusa_link_api.cpp [hunk 2]
```diff
@@ -66,32 +124,28 @@ optional<ConnectionState> PrusaLinkApi::accept(const RequestParser &parser) cons
...
+            strlcpy(putParams.filepath.data(), filename, sizeof(putParams.filepath));
             auto upload = GcodeUpload::start(parser, wui_uploaded_gcode, parser.accepts_json, std::move(putParams));
             return std::visit([](auto upload) -> ConnectionState { return std::move(upload); }, std::move(upload));
-        } else {
+        }
+        case Method::Get: {
+            return FileInfo(filename, parser.can_keep_alive(), parser.accepts_json, false);
+        }
+        case Method::Head: {
+            return FileInfo(filename, parser.can_keep_alive(), parser.accepts_json, false, true);
```
Comment (vorner) [inline lib/WUI/link_content/prusa_link_api.cpp:146]: I'd prefer to not put defaults in the parameters here; the FileInfo is called only from these two places.

Also, I'm thinking if passing the actual method (Head/Get) here would be more readable then true/false.

File: lib/WUI/link_content/prusa_link_api.cpp [hunk 3]
```diff
@@ -120,88 +174,41 @@ optional<ConnectionState> PrusaLinkApi::accept(const RequestParser &parser) cons
...
-                }
-
-                /*
-                 * Now, we make 100% sure the user won't get a file that's not
-                 * on the USB drive (eg. our xflash).
-                 */
-                if (strncmp(fname_real, "/usb/", 5) != 0) {
-                    return StatusPage(Status::Forbidden, parser.status_page_handling(), parser.accepts_json);
-                }
+            char filename[FILE_PATH_BUFFER_LEN + prefix_len];
```
Comment (vorner) [inline lib/WUI/link_content/prusa_link_api.cpp:179]: Am I right the parse_file_url outputs the filename here? Then it probably doesn't need the extra `prefix_len` any more?

I'm wondering about two things here, though.

* The buffer is not exactly small. And you now use two on the stack at the same time (the other one is inside the function). Would it make sense to declare that this parameter is not only output, but also the workspace?
* Alternatively, would it be cleaner to return some kind of `variant<array<char, ...>, error>`? I'm not sure if C++ guarantees some kind of return-placement optimization then or what, though, so maybe that could _also_ use two different buffers.

File: lib/WUI/nhttp/file_info.cpp [hunk 4]
```diff
@@ -123,7 +125,7 @@ FileInfo::FileInfo(const char *filename, bool can_keep_alive, bool json_errors,
 Step FileInfo::step(std::string_view, bool, uint8_t *output, size_t output_size) {
     const size_t last_chunk_len = strlen(LAST_CHUNK);
     size_t written = 0;
-    const ConnectionHandling handling = can_keep_alive ? ConnectionHandling::ChunkedKeep : ConnectionHandling::Close;
+    const ConnectionHandling handling = can_keep_alive && !header_only ? ConnectionHandling::ChunkedKeep : ConnectionHandling::Close;
```
Comment (vorner) [inline lib/WUI/nhttp/file_info.cpp:128]: Why do you disable keep-alive on the header_only? That probably isn't desirable.

File: lib/WUI/nhttp/file_info.cpp [hunk 5]
```diff
@@ -147,7 +149,22 @@ Step FileInfo::step(std::string_view, bool, uint8_t *output, size_t output_size)
...
+        static const char *extra_hdrs_no_printing[] = {
+            "Read-Only: false\r\n",
+            "Currently-Printing: false\r\n",
+            nullptr
+        };
+        static const char *extra_hdrs_printing[] = {
+            "Read-Only: false\r\n",
+            "Currently-Printing: true\r\n",
+            nullptr
+        };
```
Comment (vorner) [inline lib/WUI/nhttp/file_info.cpp:162]: Hmm, this doesn't look great. I think lifetime of the parameter is OK to be just during the call to `write_headers`, so you could use one static string, another buffer on stack to render the header in and have the array also on stack.

File: lib/WUI/wui_api.cpp [hunk 6]
```diff
@@ -17,6 +17,7 @@
 #include "stm32f4xx_hal.h"
 #include "print_utils.hpp"
 #include "marlin_client.hpp"
+#include "marlin_client.h"
```
Comment (vorner) [inline lib/WUI/wui_api.cpp:20]: Wait a moment. Do we have two marlin_client headers? That seems super-weird.

File: lib/WUI/wui_api.cpp [hunk 7]
```diff
@@ -404,3 +405,15 @@ bool wui_uploaded_gcode(char *filename, bool start_print) {
...
     }
 }
+
+bool wui_is_file_being_printed(const char *filename) {
+    marlin_client_loop();
+    if (!marlin_is_printing()) {
+        return false;
+    }
+
+    char SFN[FILE_PATH_BUFFER_LEN];
```
Comment (vorner) [inline lib/WUI/wui_api.cpp:415]: Style nitpick: `SFN` capitalization would look like a constant, not a variable.

File: lib/WUI/wui_api.cpp [hunk 8]
```diff
@@ -404,3 +405,15 @@ bool wui_uploaded_gcode(char *filename, bool start_print) {
...
+    strlcpy(SFN, filename, sizeof(SFN));
+    get_SFN_path(SFN);
+    return strcmp(SFN, marlin_vars()->media_SFN_path) == 0;
```
Comment (vorner) [inline lib/WUI/wui_api.cpp:418]: `strcasecmp`?

File: lib/WUI/wui_api.cpp (hunk 6 above)
Comment (HelclM) [reply lib/WUI/wui_api.cpp:20]: Apparently we do :man_shrugging: , but the `.h`  is included in `.hpp`, so the include here is redundant.

File: lib/WUI/link_content/prusa_link_api.cpp [hunk 9]
```diff
@@ -120,88 +183,41 @@ optional<ConnectionState> PrusaLinkApi::accept(const RequestParser &parser) cons
...
+            auto error = parse_file_url(parser, prefix_len, filename, FILE_PATH_BUFFER_LEN + prefix_len, RemapPolicy::Octoprint);
```
Comment (vorner) [inline lib/WUI/link_content/prusa_link_api.cpp:189]: Maybe use `sizeof(filename)` as the less error-prone alternative?

File: lib/WUI/nhttp/file_info.cpp (hunk 5 above)
Comment (vorner) [reply lib/WUI/nhttp/file_info.cpp:162]: Ping? Is this a leftover?

**Review by vorner — COMMENTED** [review]
There are few more nitpicks.

**Review by vorner — APPROVED** [review]
:+1:

**Pull request merged**